### PR TITLE
Don't add findings on hostnames with a nxdomain object

### PR DESCRIPTION
**Pull request by noamblitz**
### Changes
Dont add findings on nxdomains

### Issue link

Closes https://github.com/minvws/nl-kat-coordination/issues/1701

### Proof
_Please add some proof of your working change here, unless this is not required (e.g. this PR is trivial)._

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [ ] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: octopoes/bits/missing_dkim/missing_dkim.py [hunk 1]
```diff
@@ -1,19 +1,28 @@
...
 
-def run(input_ooi: Hostname, additional_oois: List[DKIMExists], config: Dict[str, str]) -> Iterator[OOI]:
+def run(
+    input_ooi: Hostname, additional_oois: List[Union[DKIMExists, NXDOMAIN]], config: Dict[str, str]
+) -> Iterator[OOI]:
+    dkim_exists = [ooi for ooi in additional_oois if isinstance(ooi, DKIMExists)]
+    nxdomains = [ooi for ooi in additional_oois if isinstance(ooi, NXDOMAIN)]
+
+    if nxdomains:
+        return
```
Comment (ammar92) [inline octopoes/bits/missing_dkim/missing_dkim.py:19]: Within this context is better to use a generator comprehension and use `any`, this way we don't have to go through the entire list to know if there is any `NXDOMAIN` instance within the list).
```suggestion
    nxdomains = (ooi for ooi in additional_oois if isinstance(ooi, NXDOMAIN))

    if any(nxdomains):
        return
```

File: octopoes/bits/missing_spf/missing_spf.py [hunk 2]
```diff
@@ -1,19 +1,27 @@
...
 from octopoes.models.ooi.email_security import DNSSPFRecord
 from octopoes.models.ooi.findings import Finding, KATFindingType
 
 
-def run(input_ooi: Hostname, additional_oois: List[DNSSPFRecord], config: Dict[str, str]) -> Iterator[OOI]:
+def run(
+    input_ooi: Hostname, additional_oois: List[Union[DNSSPFRecord, NXDOMAIN]], config: Dict[str, str]
+) -> Iterator[OOI]:
+    spf_records = [ooi for ooi in additional_oois if isinstance(ooi, DNSSPFRecord)]
+    nxdomains = [ooi for ooi in additional_oois if isinstance(ooi, NXDOMAIN)]
```
Comment (ammar92) [inline octopoes/bits/missing_spf/missing_spf.py:16]: See earlier comment

File: octopoes/bits/missing_dmarc/missing_dmarc.py [hunk 3]
```diff
@@ -1,19 +1,28 @@
...
 from octopoes.models.ooi.email_security import DMARCTXTRecord
 from octopoes.models.ooi.findings import Finding, KATFindingType
 
 
-def run(input_ooi: Hostname, additional_oois: List[DMARCTXTRecord], config: Dict[str, str]) -> Iterator[OOI]:
+def run(
+    input_ooi: Hostname, additional_oois: List[Union[DMARCTXTRecord, NXDOMAIN]], config: Dict[str, str]
+) -> Iterator[OOI]:
+    dmarc_records = [ooi for ooi in additional_oois if isinstance(ooi, DMARCTXTRecord)]
+    nxdomains = [ooi for ooi in additional_oois if isinstance(ooi, NXDOMAIN)]
```
Comment (ammar92) [inline octopoes/bits/missing_dmarc/missing_dmarc.py:16]: See earlier comment

File: octopoes/bits/missing_dmarc/missing_dmarc.py [hunk 4]
```diff
@@ -1,19 +1,28 @@
...
 from octopoes.models.ooi.email_security import DMARCTXTRecord
 from octopoes.models.ooi.findings import Finding, KATFindingType
 
 
-def run(input_ooi: Hostname, additional_oois: List[DMARCTXTRecord], config: Dict[str, str]) -> Iterator[OOI]:
+def run(
+    input_ooi: Hostname, additional_oois: List[Union[DMARCTXTRecord, NXDOMAIN]], config: Dict[str, str]
+) -> Iterator[OOI]:
+    dmarc_records = [ooi for ooi in additional_oois if isinstance(ooi, DMARCTXTRecord)]
+    nxdomains = (ooi for ooi in additional_oois if isinstance(ooi, NXDOMAIN))
```
Comment (underdarknl) [inline octopoes/bits/missing_dmarc/missing_dmarc.py:16]: we could return from the function in this loop immediately on the first NXDOMAIN instance.

Comment (noamblitz) [reply octopoes/bits/missing_dmarc/missing_dmarc.py:16]: Yes that's why, as per ammars request, changed it from a list to a constructor

Comment (ammar92) [reply octopoes/bits/missing_dmarc/missing_dmarc.py:16]: > Yes that's why, as per ammars request, changed it from a list to a constructor

generator* :)

**Comment by Darwinkel** [comment]
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
* No findings on `asfsdgswrgew223424.com` when it is related to an NXDOMAIN
* Findings on  `asfsdgswrgew223424.com` when it is (not yet) related to an NXDOMAIN
* Findings on `mispo.es` and other OOI's

**Review by underdarknl — APPROVED** [review]

**Pull request merged**